grudge/reductions.py

```diff
@@ -111,16 +113,19 @@ def norm(dcoll: DiscretizationCollection, vec, p, dd=None) -> float:
 
     dd = dof_desc.as_dofdesc(dd)
 
-    if isinstance(vec, np.ndarray):
+    if not isinstance(vec, DOFArray):
+        if isinstance(vec, Number):
+            return np.fabs(vec)
+
         if p == 2:
             return sum(
-                norm(dcoll, vec[idx], p, dd=dd)**2
-                for idx in np.ndindex(vec.shape)
-            )**0.5
+                norm(dcoll, comp, p, dd=dd)**2
+                for _, comp in serialize_container(vec)
+            )**(1/2)
```

Comment:
Would it make sense to use `actx.np.sqrt` here? (I changed the one in `_norm` in #170, but that PR doesn't touch this part of the code, so I didn't see this one until now.)

Reply:
I would vote to use `sqrt` where possible, but don't feel particularly strongly about it.

Reply:
Can you use `get_container_context_recursively`?

Reply:
Fair point :smile:

Reply:
> Can you use `get_container_context_recursively`?

@majosm After experimenting with this, I found a niche corner case that's mildly infuriating :laughing: here: https://github.com/inducer/grudge/blob/main/test/test_grudge.py#L1045

So it appears the original implementation (pre new world grudge) tried to be fairly general with its input. Now we're trying to assert that every `vec` being passed to `norm` has an array context. Why this isn't consistent with the other nodal reductions is likely due to an oversight on my part. (though to be honest I don't really know why we want to support the case presented [here](https://github.com/inducer/grudge/blob/main/test/test_grudge.py#L1045))

This patch makes it work consistently:
```python
if p == 2:
    from arraycontext import get_container_context_recursively

    actx = get_container_context_recursively(vec)
    if actx is not None:
        return actx.np.sqrt(
            sum(
                norm(dcoll, comp, p, dd=dd)**2
                for _, comp in serialize_container(vec)
            )
        )
    else:
        return sum(
            norm(dcoll, comp, p, dd=dd)**2
            for _, comp in serialize_container(vec)
        )**(1/2)
```

However, this special-casing is kinda ugly. I am still not solely convinced to drop `**(1/2)` when it just pollutes the code even further from my perspective (_and_ it already works as intended, no?). I am still open to being convinced.

Reply:
Fair enough. I'm wondering if we should just remove the recursion on `norm`, all this square-rooting-then-squaring seems clunky to me anyway. I might take a stab at it in a separate PR.